streams/src/test/java/org/apache/kafka/streams/state/internals/CachingWindowStoreTest.java

```diff
@@ -129,7 +129,44 @@ public void shouldPutFetchRangeFromCache() {
         assertFalse(iterator.hasNext());
         assertEquals(2, cache.size());
     }
-
+    
+    @Test
+    public void shouldGetAllFromCache() {
+        cachingStore.put(bytesKey("a"), bytesValue("a"));
+        cachingStore.put(bytesKey("b"), bytesValue("b"));
+        cachingStore.put(bytesKey("c"), bytesValue("c"));
+        cachingStore.put(bytesKey("d"), bytesValue("d"));
+        cachingStore.put(bytesKey("e"), bytesValue("e"));
+        cachingStore.put(bytesKey("f"), bytesValue("f"));
+        cachingStore.put(bytesKey("g"), bytesValue("g"));
+        cachingStore.put(bytesKey("h"), bytesValue("h"));
+
+        final KeyValueIterator<Windowed<Bytes>, byte[]> iterator = cachingStore.all();
+        String[] array = {"a", "b", "c", "d", "e", "f", "g", "h"};
+        for (String s : array) {
+            verifyWindowedKeyValue(iterator.next(), new Windowed<>(bytesKey(s), new TimeWindow(DEFAULT_TIMESTAMP, DEFAULT_TIMESTAMP + WINDOW_SIZE)), s);
+        }
+        assertFalse(iterator.hasNext());
+    }
+    
+    @Test
+    public void shouldFetchAllWithinTimestampRange() {
```

Comment:
Probably want to add a test that restricts the range, i.e, so that all values won't be removed.
You can use `context.setTime(...)` to change the timestamps, so something like;
```
context.setTime(1);
cachingStore.put(bytesKey("a"), bytesValue("a"));
context.setTime(2);
cachingStore.put(bytesKey("b"), bytesValue("b"));
context.setTime(3);
cachingStore.put(bytesKey("c"), bytesValue("c"));
context.setTime(4);
cachingStore.put(bytesKey("d"), bytesValue("d"));
...
cachingStore.fetchAll(1, 3);
```

Reply:
I'm +100 to this: we have found a few bugs on the CachingWindowStore implementation during the review which should better be caught in the test. More specifically I'd suggest even more scenarios along with @dguy 's suggestion:

1. Test that over multiple segments, the range that falls into a subset of the segments.
2. Test that within a segment, we multiple keys, each with multiple timestamps.

1) could be done in the `TimeRangeSegmentIterator` for the new class. 2) would be added here.